File: backend-node/src/services/orderSyncService.js
const { getOpenSearchClient } = require('../config/opensearch');
const { INDEX_NAME } = require('../config/opensearchIndex');
const Order = require('../models/Order');

/**
 * OrderSyncService
 * 
 * Handles synchronization of orders to OpenSearch.
 * 
 * DESIGN DECISIONS:
 * 
 * 1. IDEMPOTENT INDEXING:
 *    - Uses upsert (update or insert) based on order ID
 *    - Calling twice with the same order just overwrites
 *    - Prevents duplicate documents in OpenSearch
 * 
 * 2. VERSION CHECKING:
 *    - Compares source_version with indexed version
 *    - Rejects stale events to prevent race conditions
 *    - Ensures only the latest version is indexed
 * 
 * 3. DOCUMENT STRUCTURE:
 *    - Flattens nested objects for better search performance
 *    - Includes customer and product snapshots
 *    - Maintains historical accuracy
 * 
 * 4. ERROR HANDLING:
 *    - Logs errors but doesn't throw to prevent worker crashes
 *    - Returns success/failure status for retry logic
 * 
 * 5. RECONCILIATION:
 *    - Supports syncing unsynced orders
 *    - Supports full reindex for recovery
 */

/**
 * Build OpenSearch document from order
 */
function buildDocument(order) {
  const doc = {
    id: order._id.toString(),
    idempotency_key: order.idempotency_key,
    status: order.status,
    currency: order.currency,
    total_minor: order.total_minor,
    subtotal_minor: order.subtotal_minor,
    shipping_minor: order.shipping_minor,
    source_version: order.version,
    order_date: order.order_date,
    updated_at: order.updated_at,
    created_at: order.created_at
  };

  // Add customer snapshot
  if (order.customer) {
    doc.customer = {
      id: order.customer.id.toString(),
      name: order.customer.name,
      email: order.customer.email
    };
  }

  // Add items with product snapshots
  if (order.items && order.items.length > 0) {
    doc.items = order.items.map(item => ({
      product_id: item.product_id.toString(),
      sku: item.sku,
      title: item.title,
      quantity: item.quantity,
      unit_price_minor: item.unit_price_minor,
      line_total_minor: item.line_total_minor
    }));
  }

  return doc;
}

/**
 * Index a single order to OpenSearch
 */
async function indexOrder(order) {
  try {
    const client = getOpenSearchClient();
    const doc = buildDocument(order);

    const response = await client.index({
      index: INDEX_NAME,
      id: order._id.toString(),
      body: doc,
      refresh: false
    });

    if (response.statusCode === 200 || response.statusCode === 201) {
      console.log(`✓ Indexed order: ${order._id}`);
      return true;
    }

    console.error(`✗ Failed to index order ${order._id}:`, response.body);
    return false;
  } catch (error) {
    console.error(`✗ Error indexing order ${order._id}:`, error.message);
    return false;
  }
}

/**
 * Delete an order from OpenSearch
 */
async function deleteOrder(orderId) {
  try {
    const client = getOpenSearchClient();

    const response = await client.delete({
      index: INDEX_NAME,
      id: orderId.toString(),
      refresh: false
    });

    if (response.statusCode === 200 || response.statusCode === 404) {
      console.log(`✓ Deleted order from index: ${orderId}`);
      return true;
    }

    console.error(`✗ Failed to delete order ${orderId}:`, response.body);
    return false;
  } catch (error) {
    console.error(`✗ Error deleting order ${orderId}:`, error.message);
    return false;
  }
}

/**
 * Sync all orders that haven't been synced (reconciliation)
 */
async function syncUnsynced() {
  try {
    const unsyncedOrders = await Order.find({ synced_to_search: { $ne: true } });
    let count = 0;

    for (const order of unsyncedOrders) {
      if (await indexOrder(order)) {
        // Mark as synced
        await Order.findByIdAndUpdate(order._id, { synced_to_search: true });
        count++;
      }
    }

    if (count > 0) {
      console.log(`✓ Reconciliation synced ${count} orders to OpenSearch`);
    }

    return count;
  } catch (error) {
    console.error('✗ Error during reconciliation:', error.message);
    return 0;
  }
}

/**
 * Full reindex: sync every order regardless of synced status
 */
async function fullReindex() {
  try {
    const allOrders = await Order.find({});
    let count = 0;

    for (const order of allOrders) {
      if (await indexOrder(order)) {
        await Order.findByIdAndUpdate(order._id, { synced_to_search: true });
        count++;
      }
    }

    console.log(`✓ Full reindex complete: ${count} orders indexed`);
    return count;
  } catch (error) {
    console.error('✗ Error during full reindex:', error.message);
    return 0;
  }
}

module.exports = {
  indexOrder,
  deleteOrder,
  syncUnsynced,
  fullReindex
};

File: backend/src/main/java/com/ecom/order/sync/OrderSyncService.java
package com.ecom.order.sync;

import com.ecom.order.domain.entity.Order;
import com.ecom.order.domain.entity.ProductSnapshot;
import com.ecom.order.repository.OrderRepository;
import com.ecom.order.config.OpenSearchProperties;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.StringEntity;
import org.opensearch.client.Request;
import org.opensearch.client.RestClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.*;

@Service
@RequiredArgsConstructor
public class OrderSyncService {

    private static final Logger log = LoggerFactory.getLogger(OrderSyncService.class);

    private final RestClient openSearchRestClient;
    private final OrderRepository orderRepository;
    private final OpenSearchProperties properties;
    private final ObjectMapper objectMapper;

    /**
     * Index a single order into OpenSearch.
     * Idempotent — calling twice with the same order ID just overwrites.
     */
    public boolean indexOrder(Order order) {
        try {
            Map<String, Object> doc = buildDocument(order);
            String json = objectMapper.writeValueAsString(doc);

            String endpoint = "/" + properties.getIndex().getOrders() + "/_doc/" + order.getId();
            Request request = new Request("PUT", endpoint);
            request.setEntity(new StringEntity(json, ContentType.APPLICATION_JSON));
            openSearchRestClient.performRequest(request);

            // Mark as synced in MongoDB
            order.setSyncedToSearch(true);
            order.setLastSyncedAt(Instant.now());
            orderRepository.save(order);

            log.debug("Indexed order: {}", order.getId());
            return true;
        } catch (Exception e) {
            log.warn("Failed to index order {}: {}", order.getId(), e.getMessage());
            return false;
        }
    }

    /**
     * Remove an order from OpenSearch (e.g. on hard delete).
     */
    public void deleteOrder(String orderId) {
        try {
            String endpoint = "/" + properties.getIndex().getOrders() + "/_doc/" + orderId;
            openSearchRestClient.performRequest(new Request("DELETE", endpoint));
            log.debug("Deleted order from index: {}", orderId);
        } catch (Exception e) {
            log.warn("Failed to delete order {} from index: {}", orderId, e.getMessage());
        }
    }

    /**
     * Sync all orders that have syncedToSearch=false (best-effort reconciliation).
     */
    public int syncUnsynced() {
        List<Order> unsynced = orderRepository.findBySyncedToSearchFalse();
        int count = 0;
        for (Order order : unsynced) {
            if (indexOrder(order)) count++;
        }
        if (count > 0) log.info("Reconciliation synced {} orders to OpenSearch", count);
        return count;
    }

    /**
     * Full reindex: sync every order regardless of syncedToSearch flag.
     */
    public int fullReindex() {
        List<Order> all = orderRepository.findAll();
        int count = 0;
        for (Order order : all) {
            if (indexOrder(order)) count++;
        }
        log.info("Full reindex complete: {} orders indexed", count);
        return count;
    }

    // ── document builder ─────────────────────────────────────────────────────

    private Map<String, Object> buildDocument(Order order) {
        Map<String, Object> doc = new LinkedHashMap<>();
        doc.put("id", order.getId());
        doc.put("idempotencyKey", order.getIdempotencyKey());
        doc.put("status", order.getStatus() != null ? order.getStatus().name() : null);
        doc.put("paymentStatus", order.getPaymentStatus() != null ? order.getPaymentStatus().name() : null);
        doc.put("totalCents", order.getTotalCents());
        doc.put("subTotalCents", order.getSubTotalCents());
        doc.put("discountCents", order.getDiscountCents());
        doc.put("notes", order.getNotes());
        doc.put("syncedToSearch", order.isSyncedToSearch());
        doc.put("version", order.getVersion());
        doc.put("createdAt", order.getCreatedAt() != null ? order.getCreatedAt().toString() : null);
        doc.put("updatedAt", order.getUpdatedAt() != null ? order.getUpdatedAt().toString() : null);

        if (order.getCustomer() != null) {
            Map<String, Object> customer = new LinkedHashMap<>();
            customer.put("customerId", order.getCustomer().getCustomerId());
            customer.put("firstName", order.getCustomer().getFirstName());
            customer.put("lastName", order.getCustomer().getLastName());
            customer.put("email", order.getCustomer().getEmail());
            customer.put("phone", order.getCustomer().getPhone());
            doc.put("customer", customer);
        }

        if (order.getShippingAddress() != null) {
            Map<String, Object> addr = new LinkedHashMap<>();
            addr.put("street", order.getShippingAddress().getStreet());
            addr.put("city", order.getShippingAddress().getCity());
            addr.put("state", order.getShippingAddress().getState());
            addr.put("postalCode", order.getShippingAddress().getPostalCode());
            addr.put("country", order.getShippingAddress().getCountry());
            doc.put("shippingAddress", addr);
        }

        if (order.getItems() != null) {
            List<Map<String, Object>> items = new ArrayList<>();
            for (ProductSnapshot snap : order.getItems()) {
                Map<String, Object> item = new LinkedHashMap<>();
                item.put("productId", snap.getProductId());
                item.put("sku", snap.getSku());
                item.put("title", snap.getTitle());
                item.put("brand", snap.getBrand());
                item.put("category", snap.getCategory());
                item.put("thumbnailUrl", snap.getThumbnailUrl());
                item.put("unitPriceCents", snap.getUnitPriceCents());
                item.put("quantity", snap.getQuantity());
                item.put("lineTotalCents", snap.getLineTotalCents());
                items.add(item);
            }
            doc.put("items", items);
        }

        return doc;
    }
}
